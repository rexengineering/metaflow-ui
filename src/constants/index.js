export const BUSY = "BUSY";
export const OFFLINE = "OFFLINE";
export const ONLINE = "ONLINE";

export const ActionItemDeleteType = "ActionDeleteType";
export const ActionItemCheckType = "ActionCheckType";
export const ActionItemDisplayOnlyType = "ActionDisplayOnlyType";

export const InitialButtonState = 'initial';
export const LoadingButtonState = 'loading';
export const talkTrackIdentifierProp = 'iid';
export const talkTrackType = 'talktrack';