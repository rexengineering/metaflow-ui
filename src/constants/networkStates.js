export const INITIAL = "INITIAL";
export const REQUEST = "REQUEST";
export const SUCCESS = "SUCCESS";
export const FAILURE = "FAILURE";

export default {
  INITIAL,
  REQUEST,
  SUCCESS,
  FAILURE
}
