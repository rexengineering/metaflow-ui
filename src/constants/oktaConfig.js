const config = {
  oidc: {
    clientId: window.REX.OKTA.CLIENT_ID,
    issuer: window.REX.OKTA.ISSUER,
    redirectUri: window.REX.OKTA.REDIRECT_URI,
    scopes: ["openid", "profile", "email"],
    pkce: true,
    disableHttpsCheck: window.REX.OKTA.OKTA_TESTING_DISABLEHTTPSCHECK,
  },
  resourceServer: {
    messagesUrl: "http://localhost:5000/api/messages",
  },
};

export default config;
