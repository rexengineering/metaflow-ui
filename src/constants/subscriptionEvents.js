export const START_WORKFLOW = "START_WORKFLOW";
export const UPDATE_WORKFLOW = "UPDATE_WORKFLOW";
export const FINISH_WORKFLOW = "FINISH_WORKFLOW";
export const START_TASK = "START_TASK";
export const UPDATE_TASK = "UPDATE_TASK";
export const FINISH_TASK = "FINISH_TASK";
export const START_BROADCAST = "START_BROADCAST";
export const FINISH_BROADCAST = "FINISH_BROADCAST";
export const ERROR_BROADCAST = "ERROR_BROADCAST";
export const KEEP_ALIVE_SUCCESS = "SUCCESS";
export const KEEP_ALIVE_FAILURE = "FAILURE";