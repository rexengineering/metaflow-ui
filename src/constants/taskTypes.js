import { string, number, boolean } from "yup";

// Inputs
export const TEXT = "TEXT";
export const PHONE_NUMBER = "PHONE_NUMBER";
export const PERCENTAGE = "PERCENTAGE";
export const CURRENCY = "CURRENCY";
export const BOOLEAN = "BOOLEAN";
export const INTEGER = "INTEGER";
export const FLOAT = "FLOAT";

export function isInputType(field) {
  return [
    TEXT,
    PHONE_NUMBER,
    PERCENTAGE,
    CURRENCY,
    BOOLEAN,
    INTEGER,
    FLOAT,
  ].includes(field);
}

// Info
export const TABLE = "TABLE";
export const COPY = "COPY";
export const WORKFLOW_INSTANTIATOR = "WORKFLOW";

export function isInfoType(type) {
  return [TABLE, COPY, WORKFLOW_INSTANTIATOR].includes(type);
}

const numberValidationMessage = "Please enter a number";

export const validationSchemaMapping = {
  TEXT: string("Please enter a value"),
  PHONE_NUMBER: string("Please enter a value").length(10),
  PERCENTAGE: number(numberValidationMessage).typeError(
    numberValidationMessage
  ),
  CURRENCY: number(numberValidationMessage).typeError(numberValidationMessage),
  BOOLEAN: boolean(),
  INTEGER: number(numberValidationMessage)
    .typeError(numberValidationMessage)
    .integer(),
  FLOAT: number(numberValidationMessage).typeError(numberValidationMessage),
};

const taskTypes = {
  TEXT,
  PHONE_NUMBER,
  PERCENTAGE,
  CURRENCY,
  BOOLEAN,
  INTEGER,
  FLOAT,
  TABLE,
  COPY,
  WORKFLOW_INSTANTIATOR,
};

export default taskTypes;
