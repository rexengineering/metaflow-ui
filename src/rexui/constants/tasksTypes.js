import { string, number, boolean } from "yup";


const numberValidationMessage = "Please enter a number";

export const TEXT = "TEXT";
export const PHONE_NUMBER = "PHONE_NUMBER";
export const PERCENTAGE = "PERCENTAGE";
export const CURRENCY = "CURRENCY";
export const BOOLEAN = "BOOLEAN";
export const INTEGER = "INTEGER";
export const FLOAT = "FLOAT";
export const TABLE = "TABLE";
export const COPY = "COPY";
export const WORKFLOW_INSTANTIATOR = "WORKFLOW";

export const validationSchemaMapping = {
    TEXT: string("Please enter a value"),
    PHONE_NUMBER: string("Please enter a value").length(10),
    PERCENTAGE: number(numberValidationMessage).typeError(numberValidationMessage),
    CURRENCY: number(numberValidationMessage).typeError(numberValidationMessage),
    BOOLEAN: boolean(),
    INTEGER: number(numberValidationMessage)
        .typeError(numberValidationMessage)
        .integer(),
    FLOAT: number(numberValidationMessage).typeError(numberValidationMessage),
};

export function isInfoType(type) {
    return [TABLE, COPY, WORKFLOW_INSTANTIATOR].includes(type);
}

export function isInputType(field) {
    return [
        TEXT,
        PHONE_NUMBER,
        PERCENTAGE,
        CURRENCY,
        BOOLEAN,
        INTEGER,
        FLOAT,
    ].includes(field);
}