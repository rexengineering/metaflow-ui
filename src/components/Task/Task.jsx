import React, { useEffect, useState, useCallback } from "react";
import {
  Button,
  CircularProgress,
  makeStyles,
  Typography,
} from "@material-ui/core";
import { object } from "yup";
import PropTypes from "prop-types";
import { Formik } from "formik";
import TaskField from "../fields/TaskField";
import { completeTask } from "../../store/thunks/thunks";
import {
  buildTaskIdentifier,
  convertTaskFieldsToFormUtils,
  convertValidationErrorsTo,
} from "../../utils/tasks";
import useValidateField from "../../utils/useValidateField";
import { isInfoType, isInputType } from "../../constants/taskTypes";
import {connect} from "react-redux";
import {LoadingButtonState} from "../../constants";

const useStyles = makeStyles((theme) => ({
  form: {
    padding: theme.spacing(1),
  },
  submitButton: {
    margin: theme.spacing(8, 0, 2),
    display: "block",
    clear: "both",
    color: theme.palette.primary.contrastText,
  },
  field: {
    display: "block",
    marginTop: theme.spacing(2),
  },
  inProgressContainer: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  inProgressMessage: {
    marginLeft: theme.spacing(1),
  },
  infoField: {
    marginTop: theme.spacing(1),
  },
  circularProgress: {
    marginRight: theme.spacing(1),
  },
}));

function Task({ className, task, submitButtonText, isProcessing, isCompleted, errors, exceptionError, completeTask, isLoading }) {
  const { data } = task;
  const { formikInitialValues, validationSchema } =
    convertTaskFieldsToFormUtils(data);
  const [initialValues, setInitialValues] = useState(formikInitialValues ?? {});
  const formValidationSchema = object().shape(validationSchema);
  const onSubmit = useCallback(
    (fields) => completeTask(fields, task),
    [completeTask, task]
  );
  const validateField = useValidateField(formValidationSchema);
  const classes = useStyles();
  const { initialErrors, initialTouched } = convertValidationErrorsTo(
    errors ?? []
  );

  useEffect(() => {
    const formUtils = convertTaskFieldsToFormUtils(data);
    setInitialValues(formUtils.formikInitialValues ?? {});
  }, [data]);

  if (isCompleted) return <Typography>Talk track completed</Typography>;

  if (isProcessing || !data?.length || !initialValues) {
    return (
      <section className={classes.inProgressContainer}>
        <CircularProgress size={25} />
      </section>
    );
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmit}
      validateOnBlur={false}
      validateOnChange={false}
      enableReinitialize
      className={className}
      initialErrors={initialErrors}
      initialTouched={initialTouched}
      validationSchema={formValidationSchema}
    >
      {({ handleSubmit }) => (
        <form onSubmit={handleSubmit} className={classes.form}>
          {Array.isArray(data) &&
            data.map((field) => {
              if (isInputType(field?.type)) {
                const { dataId, label, type } = field;
                return (
                  <div key={dataId} className={classes.field}>
                    <TaskField
                      type={type}
                      id={dataId}
                      label={label}
                      validateFn={validateField}
                    />
                  </div>
                );
              }
              if (isInfoType(field?.type)) {
                const { data: fieldData, type, variant, dataId, label } = field;
                return (
                  <div className={classes.infoField} key={fieldData}>
                    <TaskField id={dataId} type={type} label={label} data={fieldData} variant={variant} />
                  </div>
                );
              }
              return <></>;
            })}

          <Button
              className={classes.submitButton}
              type="submit"
              variant="contained"
              color="primary"
              disabled={isLoading}
          >
            {
              isLoading
                  ? ( <CircularProgress className={classes.circularProgress} size={15} /> )
                  : submitButtonText
            }
          </Button>

          {exceptionError && (
            <Typography variant="body2" color="error">
              {exceptionError}
            </Typography>
          )}
        </form>
      )}
    </Formik>
  );
}

Task.defaultProps = {
  className: "",
  submitButtonText: "Submit",
  task: {},
};

Task.propTypes = {
  className: PropTypes.string,
  submitButtonText: PropTypes.string,
  task: PropTypes.shape({
    data: PropTypes.arrayOf(
      PropTypes.shape({
        data: PropTypes.string,
        dataId: PropTypes.string.isRequired,
        encrypted: PropTypes.bool.isRequired,
        label: PropTypes.string.isRequired,
        order: PropTypes.number.isRequired,
        type: PropTypes.string.isRequired,
        validators: PropTypes.arrayOf(
          PropTypes.shape({
            constraint: PropTypes.string,
            type: PropTypes.string.isRequired,
          })
        ),
      })
    ),
    iid: PropTypes.string,
    status: PropTypes.string,
    tid: PropTypes.string,
  }),
};

const mapStateToProps = (state, { task }) => {
  const { tasksState, buttons } = state.rexFlow ?? {};
  const taskIdentifier = buildTaskIdentifier(task);
  const taskState = taskIdentifier && Array.isArray(tasksState)
      ? tasksState[taskIdentifier]
      : {};
  const { isLoading, isTaskCompleted, errors: validationErrors, exceptionError: exceptions } = taskState  ?? {};
  const isProcessing = isLoading ?? false;
  const isCompleted = isTaskCompleted ?? false;
  const errors = validationErrors ?? null;
  const exceptionError = exceptions ?? null;
  return {
    isProcessing,
    isCompleted,
    errors,
    exceptionError,
    isLoading: buttons[taskIdentifier] === LoadingButtonState,
  }
};

const mapDispatchToProps = (dispatch) => ({
  completeTask: (formFields, task) => completeTask(dispatch, formFields, task)
});

export default connect(mapStateToProps, mapDispatchToProps)(Task);
