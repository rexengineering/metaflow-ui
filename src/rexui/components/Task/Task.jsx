import React, { useEffect, useState, useCallback, useContext } from "react";
import {
  Button,
  CircularProgress,
  makeStyles,
  Typography,
} from "@material-ui/core";
import { object } from "yup";
import PropTypes from "prop-types";
import { Formik } from "formik";
import {
  buildTaskIdentifier,
  convertTaskFieldsToFormUtils,
  convertValidationErrorsTo,
} from "../../utils/tasks";
import useValidateField from "../../hooks/useValidateField";
import { isInfoType, isInputType } from "../../constants/tasksTypes";
import { Context } from "../../Context";

const useStyles = makeStyles((theme) => ({
  form: {
    padding: theme.spacing(1),
  },
  submitButton: {
    margin: theme.spacing(8, 0, 2),
    display: "block",
    clear: "both",
    color: theme.palette.primary.contrastText,
  },
  field: {
    display: "block",
    marginTop: theme.spacing(2),
  },
  inProgressContainer: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  inProgressMessage: {
    marginLeft: theme.spacing(1),
  },
  infoField: {
    marginTop: theme.spacing(1),
  },
  circularProgress: {
    marginRight: theme.spacing(1),
  },
}));

/*
* ToDo: Refactor code due task state changes
* */

function Task({ task, submitButtonText, onTaskCompleted, isWorkflowFinished, className }) {

  const {
    state,
    actions: { setIsTaskCompleted },
    thunks: { completeTask },
  } = useContext(Context);

  const { data, iid } = task;
  const classes = useStyles();
  const { tasksState, activeTalkTrack, activeWorkflows } = state;
  const taskIdentifier = buildTaskIdentifier(task);
  const { formikInitialValues, validationSchema } = convertTaskFieldsToFormUtils(data);
  const formValidationSchema = object().shape(validationSchema);
  const validateField = useValidateField(formValidationSchema);

  const taskState = taskIdentifier && tasksState
      ? tasksState[taskIdentifier]
      : {};

  const customTaskState = activeTalkTrack && tasksState
      ? tasksState[activeTalkTrack]
      : {};

  const hasInvokeFinishedCallback = customTaskState?.isTaskCompleted || false;
  const { isLoading, errors: validationErrors, exceptionError: exceptions } = taskState  ?? {};
  const isProcessing = isLoading ?? false;
  const errors = validationErrors ?? null;
  const exceptionError = exceptions ?? null;

  const [initialValues, setInitialValues] = useState(formikInitialValues ?? {});
  const { initialErrors, initialTouched } = convertValidationErrorsTo( errors ?? [] );

  const currentWorkflow = Array.isArray(activeWorkflows) ? activeWorkflows.find(({iid: cuurentIid}) => cuurentIid === iid) : null;
  const onSubmit = useCallback(
      (fields) => completeTask(fields, task, currentWorkflow?.metadata?.persistent),
      [completeTask, task, currentWorkflow]
  );

  useEffect(() => {
    const formUtils = convertTaskFieldsToFormUtils(data);
    setInitialValues(formUtils.formikInitialValues ?? {});
  }, [data]);

  useEffect(() => {
    if (isWorkflowFinished && !hasInvokeFinishedCallback){
      onTaskCompleted(activeTalkTrack);
      setIsTaskCompleted(activeTalkTrack, true);
    }
  }, [isWorkflowFinished, activeTalkTrack, onTaskCompleted, hasInvokeFinishedCallback, setIsTaskCompleted]);

  if (isWorkflowFinished) {
    return <Typography>Talk track completed</Typography>;
  }

  if (isProcessing || !data?.length || !initialValues) {
    return (
      <section className={classes.inProgressContainer}>
        <CircularProgress size={25} />
      </section>
    );
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmit}
      validateOnBlur={false}
      validateOnChange={false}
      enableReinitialize
      className={className}
      initialErrors={initialErrors}
      initialTouched={initialTouched}
      validationSchema={formValidationSchema}
    >
      {({ handleSubmit }) => (
        <form onSubmit={handleSubmit} className={classes.form}>
          {Array.isArray(data) &&
            data.map((field) => {
              if (isInputType(field?.type)) {
                const { dataId, label, type } = field;
                return (
                  <div key={dataId} className={classes.field}>
                    <TaskField
                      type={type}
                      id={dataId}
                      label={label}
                      validateFn={validateField}
                    />
                  </div>
                );
              }
              if (isInfoType(field?.type)) {
                const { data: fieldData, type, variant, dataId, label } = field;
                return (
                  <div className={classes.infoField} key={fieldData}>
                    <TaskField id={dataId} type={type} label={label} data={fieldData} variant={variant} />
                  </div>
                );
              }
              return <></>;
            })}

          <Button
              className={classes.submitButton}
              type="submit"
              variant="contained"
              color="primary"
              disabled={isLoading}
          >
            {
              isLoading
                  ? ( <CircularProgress className={classes.circularProgress} size={15} /> )
                  : submitButtonText
            }
          </Button>

          {exceptionError && (
            <Typography variant="body2" color="error">
              {exceptionError}
            </Typography>
          )}
        </form>
      )}
    </Formik>
  );
}

Task.defaultProps = {
  className: "",
  submitButtonText: "Submit",
  task: {},
  onTaskCompleted: () => {},
  isWorkflowFinished: false,
};

Task.propTypes = {
  className: PropTypes.string,
  submitButtonText: PropTypes.string,
  isWorkflowFinished: PropTypes.bool,
  task: PropTypes.shape({
    data: PropTypes.arrayOf(
      PropTypes.shape({
        data: PropTypes.string,
        dataId: PropTypes.string.isRequired,
        encrypted: PropTypes.bool.isRequired,
        label: PropTypes.string.isRequired,
        order: PropTypes.number.isRequired,
        type: PropTypes.string.isRequired,
        validators: PropTypes.arrayOf(
          PropTypes.shape({
            constraint: PropTypes.string,
            type: PropTypes.string.isRequired,
          })
        ),
      })
    ),
    iid: PropTypes.string,
    status: PropTypes.string,
    tid: PropTypes.string,
  }),
  onTaskCompleted: PropTypes.func,
};


export default Task;
