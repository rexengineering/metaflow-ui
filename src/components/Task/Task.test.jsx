import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import Task from "./Task";
import getStore from "../../store";
import { TEXT, COPY } from "../../constants/taskTypes";

describe("<Task />", () => {
  const props = {
    className: "",
    task: {
      iid: "SDT32ER64W",
      status: "UP",
      tid: "TERWT65756",
      data: [
        {
          data: "Gabriel",
          dataId: "name",
          encrypted: false,
          label: "Name",
          order: 1,
          type: TEXT,
          validators: [
            {
              constraint: "",
              type: "",
            },
          ],
        },
        {
          data: "some copy",
          dataId: "copy",
          encrypted: false,
          label: "Name",
          order: 2,
          type: COPY,
        },
      ],
    },
  };
  const renderTask = ({ className, task } = props, store = getStore({})) => (
    <Provider store={store}>
      <Task className={className} task={task} />
    </Provider>
  );

  it("should render", () => {
    const { asFragment } = render(renderTask());
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(renderTask());
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should render a submit button", () => {
    render(renderTask());
    const buttonNode = screen.getByText("Submit");
    expect(buttonNode).toBeInTheDocument();
  });

  it("should render copy text", () => {
    render(renderTask());
    const copyNode = screen.getByText("some copy");
    expect(copyNode).toBeInTheDocument();
  });
});
