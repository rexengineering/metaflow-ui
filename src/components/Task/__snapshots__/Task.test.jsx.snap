// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Task /> should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="makeStyles-form-7"
  >
    <div
      class="makeStyles-field-9"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink="true"
        >
          Name
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiFilledInput-input"
            name="name"
            type="text"
            value="Gabriel"
          />
        </div>
      </div>
    </div>
    <div
      class="makeStyles-infoField-12"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        some copy
      </p>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submitButton-8 MuiButton-containedPrimary"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
